### DISCUSSION TITLE
Where's the UI code?

### DISCUSSION THREAD

**Discussion by sli** [PlummersSoftwareLLC/NightDriverStrip#208] [General]
In the recent [laser cannon](https://www.youtube.com/watch?v=1hYmpztMQe0) video, Dave mentioned wanting contributors for the React UI. However, the repo seems to only contain a compiled bundle and no original source for the UI side. I'd like to contribute to that and also look into writing alternatives that minimize the bundle size. Is there another repo somewhere that I just can't find?

**Comment by davepl** [comment]
Someone has already started on it, so they’re working in a separate repo.  There’s lot to do, however, and I’d be happy to connect you two fellows!  If you want to send me your email, I’ll pass it along and you can coordinate in the same fork, ideally.  

Let me know if you’re interested, as two brains are better than one!

- Dave

**Reply by monomalvado** [reply]
I am also interested in contributing to the React UI for this project.

**Comment by davepl** [comment]
Cool!  I’ve added Louis Riel, who is currently working on the UI in React; please see if you can coordinate with him!  He has a project up and running, and has a source fork already.   I’m sure there’s lots of work to go around.

Right now I’m holding up things since I haven’t provided layout pics yet, but I finished this week’s episode and will work on that tomorrow.

Please let me know if I can help work out any kinks or help coordinate!

Cheers,
Dave

**Reply by Louis-Riel** [reply]
If you want to take a look at the site, go see my fork in the site folder over there.

It is in the very early stages, although it works and has all of the API calls implemented. As of now, the site does not actually compile per say, it just concatenates the source files together, and uses the browsers babel compiler. We'd need to see if we want to keep it that way, or actually compile it. The site currently stands at 64k, and is now part of the main flash. If it being compiled would grow it, then maybe we don't do that, and continue use the browser's babel pre compiler, throwing this "this is not appropriate for production site" browser warning to the wind.

I have not yet gotten my LEDs, so I'm running blind, and as of yesterdays master, my chip would reboot on API calls a bunch, maybe your results would vary.

For the work remaining, we have
- Layout, especially considerations for mobile screens. 
- Solidifying the theming (got a few colors explicitly specified in there)
- Light mode to the current dark mode. 

Please note that the site is built to allow for local development, follow the instructions over my fork to get that setup working. 

The site is currently built to be runnable as a self served site by the chip talking back to it's self, or have the site be hosted outside the chip, in a hosted mode, with the IP of the chip configured. I want the code of the site to remain this way for future considerations of the bright future where folks are running a bunch of chips and lights, making their house visible from space. Having the site this way would allow the site to point to any chip, were you to run multiples like Dave. That way, you could either:
- point your browser to any chip, and from there, select any other chip, assuming we have that picklist built.
- Go all out and host the same site on NGINX and put in security tooling to prevent anybody on your LAN to mess with your lights.
- Go out even further with NGINX and implement a security model to address the IOT considerations so many don't address. I got a model designed for this I call NIOT, maybe I'll use this to demonstrate it. This is probably outside the scope of this work, but I would like to keep the code modularized properly to allow for that.

Future work could include integrating with the chips socket through a websocket implementation for real time updates and interactions.

Thank, Louis.
